torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,1737 @@
+            # `n != ""` excludes self.
+            if n != "" and isinstance(m, FullyShardedDataParallel):
+                # We relax the assert for non-root instance, when the nested inialized module is wrapped
+                # again in FSDP later, for example after training to run inference.
+                assert m._is_root is None or not m._is_root
+                if m._is_root is None:
+                    m._is_root = False
+                if m.process_group != self.process_group:
+                    self.children_share_process_group = False
+
+                # if child instance in its own (smaller) world, that was probably an attempt to avoid OOM.
+                # Therefore gathering this child's optim state will probably cause OOM, so we won't do it.
+                m.no_broadcast_optim_state = m.no_broadcast_optim_state or (
+                    (m.world_size == 1) and (m.world_size < self.world_size) and (m.process_group != self.process_group)
+                )
+
+    def _setup_streams(self) -> None:
+        """Create streams to overlap data transfer and computation."""
+        if len(self._streams) > 0 or not self._is_root:
+            return
+
+        if torch.cuda.is_available():
+            # Stream to move main FP32 params (may be on CPU) to FP16 for forward.
+            self._streams["fp32_to_fp16"] = torch.cuda.Stream()
+            # Stream for all-gathering parameters.
+            self._streams["all_gather"] = torch.cuda.Stream()
+            # Stream for overlapping grad reduction with the backward pass.
+            self._streams["post_backward"] = torch.cuda.Stream()
+
+        # Helper for bucketing reduce-scatter ops. This is also shared with
+        # children instances to improve bucket utilization.
+        self._reducer = ReduceScatterBucketer(self.bucket_cap_mb)
+        # We share streams with all children instances, which allows them to
+        # overlap transfers across the forward pass without synchronizing with
+        # the default stream.
+        for n, m in self.named_modules():
+            if n != "" and isinstance(m, FullyShardedDataParallel):
+                m._streams = self._streams
+                m._reducer = self._reducer
+
+    def _wait_for_previous_optim_step(self) -> None:
+        """
+        The outer-most :class:`FullyShardedDataParallel` instance (i.e., the root
+        instance) needs to synchronize with the default stream to ensure the
+        previous optimizer step is done.
+        """
+        if not torch.cuda.is_available():
+            return
+        if self.mixed_precision:
+            self._streams["fp32_to_fp16"].wait_stream(torch.cuda.current_stream())
+        else:
+            self._streams["all_gather"].wait_stream(torch.cuda.current_stream())
+
+    def forward(self, *args: Any, **kwargs: Any) -> torch.Tensor:
+        self._lazy_init()
+
+        # Start of a forward pass.
+        self.training_state = TrainingState.FORWARD
```

Comment:
just a topic for discussion:

These states might make wrap FSPD with Pipeline (or the other way around) hard, because different micro-batches might be in different stages. But I think this is fine, as FSDP can do prefetching itself to overlap comm with comp. Besides, FSDP + Pipeline probably is not the reasonable marriage, as pipeline will need to offload-materialize each layer multiple times.

Reply:
I do not know much about pipeline details:(.  these states are mainly used to validate whether previous stages are executed as expected (cleaner than using flags in DDP). I guess it will not be too hard to add or refactor these states when we want to support FSDP with other modules

Reply:
The reason that I think Pipeline and FSDP does not work well together is because FSDP needs expensive allgather in each forward and backward passes to get the parameters. If we combine FSDP with pipelining, then each micro-batch will result in an additional allgather, which could make it super slow and overshadow the benefits of enabling concurrently computes on different devices.